Minor fixes and update cts

kunalmohan opened the pull request:
<!-- Please describe your changes on the following line: -->
- Prevent redundant buffer and texture destroy calls.
- More subtests for B2B copy pass now.
- All tests under `setViewport()` and `setScissorRect()` pass now.
- Tests for `createTexture()` do not crash. More than 50% of them pass now.

r?@kvark


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

kunalmohan committed:
Prevent redundant texture and buffer destroy calls

kunalmohan committed:
update cts and test expectations

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/gpubuffer.rs, components/script/dom/gputexture.rs
 * @jgraham: tests/wpt/webgpu/tests/webgpu/common/framework/params_utils.js, tests/wpt/webgpu/tests/webgpu/cts.html, tests/wpt/webgpu/tests/webgpu/webgpu/util/texture/layout.js, tests/wpt/webgpu/tests/webgpu/webgpu/api/validation/setScissorRect.spec.js, tests/wpt/webgpu/meta/webgpu/cts.html.ini and 8 more
 * @kichjang: components/script/dom/gpubuffer.rs, components/script/dom/gputexture.rs

kunalmohan commented:
@bors-servo try=wpt-mac

bors-servo commented:
:hourglass: Trying commit 49d3012584aec9f324577a2dcc6918ab439e96d1 with merge 4ffbf2daa74a34bfeb1ad5a6e36eaf051f6134d4...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/O9_hpY_wREyDD-soWSIDQQ)

kunalmohan commented:
```
 ▶ CRASH [expected OK] /_webgpu/webgpu/cts.html?q=webgpu:api,validation,createTexture:*
  │ 
  │ _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
  │ -[MTLTextureDescriptorInternal validateWithDevice:]:842: failed assertion `MTLTextureDescriptor sampleCount (2) is not supported by device.'
  │ Stack trace for thread "WGPU"
  │    0: <servo::backtrace::Print as core::fmt::Debug>::fmt
  │    1: core::fmt::write
  │    2: std::io::Write::write_fmt
  │    3: servo::backtrace::print
  │    4: servo::install_crash_handler::handler
  └    5: __sigtramp
```
Unexpected crash on metal. Reason- validation for `createTexture()` is not complete.
I'm going to disable these tests for now.

kunalmohan commented:
```
STDERR: Exception in thread ProcessReader:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/worker/tasks/task_1598336377/repo/python/_virtualenv2.7/lib/python2.7/site-packages/mozprocess/processhandler.py", line 1116, in _read
    callback(line.rstrip())
  File "/Users/worker/tasks/task_1598336377/repo/python/_virtualenv2.7/lib/python2.7/site-packages/mozprocess/processhandler.py", line 1026, in __call__
    e(*args, **kwargs)
  File "/Users/worker/tasks/task_1598336377/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/executorservo.py", line 152, in on_output
    self.result_data = json.loads(line[len(prefix):])
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 382, in raw_decode
    obj, end = self.scan_once(s, idx)
ValueError: Expecting object: line 1 column 1737729 (char 1737728)
```
@jdm Any thoughts on this? It occurs in my local system too.

kunalmohan committed:
disable createTexture() validation tests

kunalmohan commented:
@bors-servo try=wpt-mac

bors-servo commented:
:hourglass: Trying commit d3d006dea3c3eb249d7a9beabbe6c06c7203a29f with merge e7a08791fb79134ded1563a70854341fce2d1b6f...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/I7eTFgv6Qrm5wPlH6PZBGg)
State: approved= try=True

kunalmohan commented:
> ```
> STDERR: Exception in thread ProcessReader:
> Traceback (most recent call last):
>   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
>     self.run()
>   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
>     self.__target(*self.__args, **self.__kwargs)
>   File "/Users/worker/tasks/task_1598336377/repo/python/_virtualenv2.7/lib/python2.7/site-packages/mozprocess/processhandler.py", line 1116, in _read
>     callback(line.rstrip())
>   File "/Users/worker/tasks/task_1598336377/repo/python/_virtualenv2.7/lib/python2.7/site-packages/mozprocess/processhandler.py", line 1026, in __call__
>     e(*args, **kwargs)
>   File "/Users/worker/tasks/task_1598336377/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/executorservo.py", line 152, in on_output
>     self.result_data = json.loads(line[len(prefix):])
>   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 338, in loads
>     return _default_decoder.decode(s)
>   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 366, in decode
>     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
>   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 382, in raw_decode
>     obj, end = self.scan_once(s, idx)
> ValueError: Expecting object: line 1 column 1737729 (char 1737728)
> ```

This is still there.

jdm commented:
I recommend modifying the source from:
File "/Users/worker/tasks/task_1598336377/repo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/executorservo.py", line 152, in on_output
    self.result_data = json.loads(line[len(prefix):])

To add a debug print that shows you the json being loaded.

kunalmohan commented:
How do I enable logs while running wpt? I don't see them in my terminal. Any print statements in the rust code or python code don't show up on my terminal while running wpt.

jdm commented:
Run the test-wpt command with `--log-mach -`.

kvark commented on components/webgpu/lib.rs:
```diff
@@ -475,9 +474,7 @@ impl<'a> WGPU<'a> {
-                        if result.is_err() {
-                            self.encoder_record_error(command_encoder_id, result.clone());
-                        }
+                        self.encoder_record_error(command_encoder_id, result.clone());
```
instead of cloning the error unconditionally, we should probably pass it by reference and clone inside the function only if needed

kunalmohan committed:
prevent unconditional cloning of results

kunalmohan commented:
Every time I try to run wpt with the logging on, I get timeouts, and, therefore, I am unsuccessful in investigating the `ValueError` in wptrunner. Also, I've been a bit busy for the past two days so couldn't devote much time to it.

 I have addressed the review comments, and since the error does not give us unexpected test results, I hope that we can merge this now (Since this PR includes a number of tests, I plan to get it merged before the official timeline of the project ends). I'll try to investigate the issue further once I get some time.

kvark commented:
Is there a way to bump the timeouts?
Anyhow, let's proceed
@bors-servo r+

bors-servo commented:
:pushpin: Commit 85b6bbb has been approved by `kvark`

<!-- @bors-servo r=kvark 85b6bbb33ace8b433ed279d2cb73ed8d96bb9690 -->

bors-servo commented:
:hourglass: Testing commit 85b6bbb33ace8b433ed279d2cb73ed8d96bb9690 with merge 11ba84353013ce30ea2fdff73697238c761583f4...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Kr1MqJYpRT-AEFWHLx85sg)

jdm commented:
@bors-servo retry
* #25513 #25522

bors-servo commented:
:hourglass: Testing commit 85b6bbb33ace8b433ed279d2cb73ed8d96bb9690 with merge 9e6da58d7793a4576fef38446457e1073a19cd5e...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/I51SEgUXSma7fSMUBmL7-Q)
Approved by: kvark
Pushing 9e6da58d7793a4576fef38446457e1073a19cd5e to master...
